lib/enkf/row_scaling.cpp

```diff
@@ -0,0 +1,208 @@
+/*
+   Copyright (C) 2020  Equinor ASA, Norway.
+
+   The file 'row_scaling.cpp' is part of ERT - Ensemble based Reservoir Tool.
+
+   ERT is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   ERT is distributed in the hope that it will be useful, but WITHOUT ANY
+   WARRANTY; without even the implied warranty of MERCHANTABILITY or
+   FITNESS FOR A PARTICULAR PURPOSE.
+
+   See the GNU General Public License at <http://www.gnu.org/licenses/gpl.html>
+   for more details.
+*/
+#include <algorithm>
+#include <cmath>
+#include <numeric>
+#include <stdexcept>
+#include <vector>
+
+
+#include <ert/enkf/row_scaling.hpp>
+#include <ert/res_util/matrix.hpp>
+#include <ert/util/util.hpp>
+
+#define ROW_SCALING_ID 6123199
+#define ROW_SCALING_RESOLUTION 1000
+
+
+/*
+  The values in the row_scaling container are distributed among
+  ROW_SCALING_RESOLUTION discreete values. The reason for this lumping is to be
+  able to reuse the scaled update matrix:
+
+      X(a) = a*X0 + (1 - a)*I
+
+  for several rows. This grouping is utilized in the row_scaling_multiply()
+  function.
+*/
+
+
+struct row_scaling_struct {
+  UTIL_TYPE_ID_DECLARATION;
```

Comment:
Well - the `local_dataset_type` has `hash_type` container of these objects, they need to be castable from `void *` to register a destructor in the `hash` container, and then this "type system" ensures some safety in the casting. Again - I will not argue the virtues of this system, but it has been quite a lot used in `libecl / libres`. The `xxx_safe_cast()` functions seldom trigger - but when they do they give the developer a quite good direction of where the bug is (of course - while casting from `void *` the alternative is silent "success").

Reply:
I will not push hard against this, but I think it's nice to consider alternatives so that we don't add more technical debt. I can try to solve it if you don't have time.